File: backEnd/spaced_repetition.py
"""Spaced repetition (SM-2) utilities for LockIN backend.

This module provides a pure implementation of the SM-2 algorithm and
helpers to compute card scheduling fields. It does not depend on a DB
package so it can be used in unit tests and in your backend code.

Functions:
- `sm2_update(card, quality, now=None)` -> returns updated fields dict

Card argument is a mapping with at least these keys:
  - `ease_factor` (float)
  - `repetitions` (int)
  - `interval_days` (int)

`quality` follows Anki's 0-5 scale (int).
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Dict, Any, Optional, TYPE_CHECKING


if TYPE_CHECKING:
    import psycopg2


@dataclass
class CardState:
    ease_factor: float
    repetitions: int
    interval_days: int


def _now_utc() -> datetime:
    return datetime.now(timezone.utc)


def sm2_update(card: Dict[str, Any], quality: int, now: Optional[datetime] = None) -> Dict[str, Any]:
    """Compute SM-2 updates for a card given a `quality` (0-5).

    Args:
      card: mapping containing `ease_factor` (float), `repetitions` (int),
            and `interval_days` (int). Other keys are ignored.
      quality: integer 0..5 (Anki-style quality rating)
      now: optional datetime for `last_reviewed` and `due_date` computation

    Returns:
      dict with updated keys: `ease_factor`, `repetitions`, `interval_days`,
      `due_date` (ISO str), `last_reviewed` (ISO str)

    Notes:
      - Uses classic SM-2 formula for ease factor adjustment and intervals.
      - Ensures ease factor >= 1.3.
    """
    if now is None:
        now = _now_utc()

    # Defensive defaults
    ef = float(card.get("ease_factor", 2.5))
    reps = int(card.get("repetitions", 0))
    prev_interval = int(card.get("interval_days", 0))

    # the clamp quality
    q = int(quality)
    if q < 0:
        q = 0
    if q > 5:
        q = 5

    if q < 3:
        # the failed review, which will then reset repetitions
        reps = 0
        interval = 1
    else:
        reps += 1
        if reps == 1:
            interval = 1
        elif reps == 2:
            interval = 6
        else:
            # for repetitions > 2 use previous interval * EF
            # if prev_interval is 0 (defensive), treat it as 6
            base = prev_interval if prev_interval > 0 else 6
            interval = round(base * ef)

    # update ease factor
    new_ef = ef + 0.1 - (5 - q) * (0.08 + (5 - q) * 0.02)
    if new_ef < 1.3:
        new_ef = 1.3

    due_date = now + timedelta(days=interval)

    return {
        "ease_factor": round(new_ef, 4),
        "repetitions": reps,
        "interval_days": int(interval),
        "due_date": due_date.isoformat(),
        "last_reviewed": now.isoformat(),
    }


def update_card_db(conn, card_id: int, quality: int, now: Optional[datetime] = None) -> Dict[str, Any]:
    """Update a card row in the DB using a DB-API connection.

    This function expects `conn` to be a DB-API connection (e.g. psycopg2).
    It will SELECT the current card fields, compute updates, apply them
    with an UPDATE, and return the new fields.

    If `psycopg2` is not installed or the connection doesn't match the
    expected interface, the function will raise an informative error.
    """
    try:
        import psycopg2
    except Exception as exc:
        raise RuntimeError("psycopg2 is required for DB updates: install psycopg2-binary") from exc

    cur = conn.cursor()
    cur.execute(
        "SELECT ease_factor, repetitions, interval_days FROM \"Card\" WHERE id = %s",
        (card_id,),
    )
    row = cur.fetchone()
    if not row:
        cur.close()
        raise ValueError(f"Card id={card_id} not found")

    card = {"ease_factor": float(row[0]), "repetitions": int(row[1]), "interval_days": int(row[2])}
    updates = sm2_update(card, quality, now=now)

    cur.execute(
        "UPDATE \"Card\" SET ease_factor = %s, repetitions = %s, interval_days = %s, due_date = %s, last_reviewed = %s WHERE id = %s RETURNING id",
        (
            updates["ease_factor"],
            updates["repetitions"],
            updates["interval_days"],
            updates["due_date"],
            updates["last_reviewed"],
            card_id,
        ),
    )
    conn.commit()
    cur.close()
    return updates


if __name__ == "__main__":
    # demo
    sample = {"ease_factor": 2.5, "repetitions": 0, "interval_days": 0}
    print("Initial:", sample)
    for q in [5, 5, 5, 4, 2, 5]:
        updates = sm2_update(sample, q)
        print(f"quality={q} ->", updates)
        # apply updates to sample for next iteration
        sample["ease_factor"] = updates["ease_factor"]
        sample["repetitions"] = updates["repetitions"]
        sample["interval_days"] = updates["interval_days"]

File: backEnd/spaced_repetition.js
/*
 * Spaced repetition (SM-2) utilities for LockIN backend (JavaScript).
 *
 * Exports:
 * - sm2Update(card, quality, now?) -> updated fields object
 * - updateCardDb(client, cardId, quality, now?) -> Promise resolving to updates
 */
'use strict';

function _nowUtc() {
  return new Date();
}

function clampInt(v, lo, hi) {
  v = Number(v) | 0;
  if (v < lo) return lo;
  if (v > hi) return hi;
  return v;
}

/**
 * Compute SM-2 updates for a card.
 * card: object with ease_factor (number), repetitions (int), interval_days (int)
 * quality: integer 0..5
 * now: optional Date
 * Returns object with keys: ease_factor, repetitions, interval_days, due_date, last_reviewed
 */
function sm2Update(card = {}, quality, now = null) {
  if (!now) now = _nowUtc();

  const ef = Number(card.ease_factor !== undefined ? card.ease_factor : 2.5);
  let reps = Number.isFinite(card.repetitions) ? (card.repetitions | 0) : 0;
  const prev_interval = Number.isFinite(card.interval_days) ? (card.interval_days | 0) : 0;

  let q = clampInt(quality, 0, 5);

  let interval;
  if (q < 3) {
    reps = 0;
    interval = 1;
  } else {
    reps += 1;
    if (reps === 1) {
      interval = 1;
    } else if (reps === 2) {
      interval = 6;
    } else {
      const base = prev_interval > 0 ? prev_interval : 6;
      interval = Math.round(base * ef);
    }
  }

  let newEf = ef + 0.1 - (5 - q) * (0.08 + (5 - q) * 0.02);
  if (newEf < 1.3) newEf = 1.3;

  const dueDate = new Date(now.getTime() + interval * 24 * 60 * 60 * 1000);

  return {
    ease_factor: Math.round(newEf * 10000) / 10000,
    repetitions: reps,
    interval_days: Number(interval),
    due_date: dueDate.toISOString(),
    last_reviewed: now.toISOString(),
  };
}

/**
 * Update a card row in the DB using a node-postgres client or pool.
 * client: any object with `query(text, params)` returning a Promise.
 * Returns the computed updates.
 */
async function updateCardDb(client, cardId, quality, now = null) {
  if (!client || typeof client.query !== 'function') {
    throw new Error('A node-postgres client (with query()) is required for DB updates');
  }
  const res = await client.query('SELECT ease_factor, repetitions, interval_days FROM "Card" WHERE id = $1', [cardId]);
  if (!res || !res.rows || res.rows.length === 0) {
    throw new Error(`Card id=${cardId} not found`);
  }
  const row = res.rows[0];
  const card = { ease_factor: Number(row.ease_factor), repetitions: Number(row.repetitions), interval_days: Number(row.interval_days) };

  const updates = sm2Update(card, quality, now);

  await client.query(
    'UPDATE "Card" SET ease_factor = $1, repetitions = $2, interval_days = $3, due_date = $4, last_reviewed = $5 WHERE id = $6',
    [updates.ease_factor, updates.repetitions, updates.interval_days, updates.due_date, updates.last_reviewed, cardId]
  );

  return updates;
}

module.exports = { sm2Update, updateCardDb };

if (require.main === module) {
  // demo
  let sample = { ease_factor: 2.5, repetitions: 0, interval_days: 0 };
  console.log('Initial:', sample);
  for (const q of [5, 5, 5, 4, 2, 5]) {
    const updates = sm2Update(sample, q);
    console.log(`quality=${q} ->`, updates);
    sample.ease_factor = updates.ease_factor;
    sample.repetitions = updates.repetitions;
    sample.interval_days = updates.interval_days;
  }
}
